filament/backend/src/vulkan/memory/ResourceManager.h

```diff
@@ -0,0 +1,149 @@
+        auto s = sSingleton;
+        s->mHandleAllocatorImpl.associateTagToHandle(id, std::move(tag));
+    }
+
+    static inline void print() {
+        auto s = sSingleton;
+        s->printImpl();
+    }
+
+    static inline void gc() {
+        auto s = sSingleton;
+        s->gcImpl();
+    }
+
+private:
+    using AllocatorImpl = HandleAllocatorVK;
+
+    template <typename D>
+    using requires_thread_safety = typename std::disjunction<
+        std::is_same<D, VulkanFence>, std::is_same<D, VulkanTimerQuery>>;
+
+    ResourceManager(size_t arenaSize, bool disableUseAfterFreeCheck);
+
+    template<typename D, typename B, typename... ARGS>
+    static inline RefCounter construct(Handle<B> const& handle, ARGS&&... args) noexcept {
+        // TODO: to be implemented by uncommenting
+        // constexpr bool THREAD_SAFETY = requires_thread_safety<D>::value;
+        // auto s = sSingleton;
+        // D* obj = s->mHandleAllocatorImpl.construct<D, B>(handle, std::forward<ARGS>(args)...);
+        // obj->counterIndex = s->mPool.create<THREAD_SAFETY>(getTypeEnum<D>(), handle.getId());
+        // s->trackIncrement(getTypeEnum<D>(), handle.getId());
+        // return s->mPool.get<THREAD_SAFETY>(obj->counterIndex);
+
+        return RefCounter();
+    }
+
+    template<typename Dp, typename B>
+    static inline typename std::enable_if_t<
+            std::is_pointer_v<Dp> && std::is_base_of_v<B, typename std::remove_pointer_t<Dp>>, Dp>
+    handle_cast(Handle<B> const& handle) noexcept {
+        auto s = sSingleton;
+        return s->mHandleAllocatorImpl.handle_cast<Dp, B>(handle);
+    }
+
+    template<typename D, typename B>
+    static inline void destruct(Handle<B> handle) noexcept {
+        // TODO: to be implemented by uncommenting
+        // constexpr bool THREAD_SAFETY = requires_thread_safety<D>::value;
+        // auto s = sSingleton;
+        // D* obj = s->handle_cast<D*>(handle);
+        // CounterIndex const counterIndex = obj->counterIndex;
+        // s->mPool.free<THREAD_SAFETY>(counterIndex);
+        // s->mHandleAllocatorImpl.deallocate(handle, obj);
+    }
+
+    template<typename D>
+    static RefCounter fromCounterIndex(CounterIndex counterInd) {
+        constexpr bool THREAD_SAFETY = requires_thread_safety<D>::value;
+        auto s = sSingleton;
+        return s->mPool.get<THREAD_SAFETY>(counterInd);
+    }
+
+    static void destroyWithType(ResourceType type, HandleId id);
+
+    // This will post the destruction to a list that will be processed when gc is called from the
+    // backend thread.
+    inline static void destroyLaterWithType(ResourceType type, HandleId id) {
+        auto s = sSingleton;
+        std::unique_lock<utils::Mutex> lock(s->mGcListMutex);
+        s->mGcList.push_back({type, id});
+    }
+
+    void gcImpl();
+    void printImpl() const;
+    void trackIncrement(ResourceType type, HandleId id);
+
+    AllocatorImpl mHandleAllocatorImpl;
+    static ResourceManager* sSingleton;
```

Comment:
is this really what you want? What if a process has to instances of filament? Do you want them to share this? what would the benefit be?

Reply:
removed singleton